plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1363 @@
+            raise TypeError(f"Invalid tolerance: {value}") from exc
+        if not 0 <= value < 1:
+            raise ValueError("Need 0 <= tol < 1.")
+        self._tol = value
+
+    @property
+    def time_input(self) -> u.s:
+        return self._time_input
+
+    @time_input.setter
+    def time_input(self, times: u.s):
+        if times is None:
+            self._time_input = None
+        elif isinstance(times, u.Quantity):
+            if times.isscalar:
+                raise ValueError("time_input must be an array.")
+            try:
+                times = times.to(u.s)
+            except u.UnitConversionError:
+                raise u.UnitsError("time_input must have units of seconds.") from None
+            if not np.all(times[1:] > times[:-1]):
+                raise ValueError("time_input must monotonically increase.")
+            self._time_input = times
+        else:
+            raise TypeError("Invalid time_input.")
+
+    @property
+    def time_start(self) -> u.s:
+        """The start time of the simulation."""
+        return self._time_start
+
+    @time_start.setter
+    def time_start(self, time: u.s):
+        if time is None:
+            self._time_start = 0.0 * u.s
+        elif isinstance(time, u.Quantity):
+            if not time.isscalar:
+                raise ValueError("time_start must be a scalar")
+            try:
+                time = time.to(u.s)
+            except u.UnitConversionError:
+                raise u.UnitsError("time_start must have units of seconds") from None
+            if (
+                hasattr(self, "_time_max")
+                and self._time_max is not None
+                and self._time_max <= time
+            ):
+                raise ValueError("Need time_start < time_max.")
+            if self.time_input is not None and self.time_input.min() > time:
+                raise ValueError("time_start must be less than min(time_input)")
+            self._time_start = time
+        else:
+            raise TypeError("Invalid time_start.") from None
+
+    @property
+    def time_max(self) -> u.s:
+        """The maximum time allowed for the simulation."""
+        return self._time_max
+
+    @time_max.setter
+    def time_max(self, time: u.s):
+        if time is None:
+            self._time_max = (
+                self.time_input[-1] if self.time_input is not None else np.inf * u.s
+            )
+        elif isinstance(time, u.Quantity):
+            if not time.isscalar:
+                raise ValueError("time_max must be a scalar")
+            try:
+                time = time.to(u.s)
+            except u.UnitConversionError:
+                raise u.UnitsError("time_max must have units of seconds") from None
+            if (
+                hasattr(self, "_time_start")
+                and self._time_start is not None
+                and self._time_start >= time
+            ):
+                raise ValueError("time_max must be greater than time_start")
```

Comment:
The two times have a codependence on each other, but I still wonder if it might not make sense to extract this code out?